File: js/102_BinaryTreeLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

 //recursive
var levelOrder = function(root) {
    var ret = [];
    buildVector(root,0,ret);
    return ret;
};

var buildVector = function(root,depth,ret){
    if(root === null) return;
    if(depth === ret.length) ret.push([]);
    ret[depth].push(root.val);
    buildVector(root.left,depth+1,ret);
    buildVector(root.right,depth+1,ret);
};

//iterative
var levelOrder = function(root) {
    var ret = [];
    if(root === null) return ret;
    
    var queue = [];
    queue.push(root);
    queue.push(null);
    
    var cur = [];
    
    while(queue.length > 0){
        var t = queue.shift();
        if( t === null ){
            ret.push(cur);
            cur = [];
            if(queue.length) queue.push(null);
        }else{
            cur.push(t.val);
            if(t.left !== null) queue.push(t.left);
            if(t.right !== null) queue.push(t.right);
        }
    }
    return ret;
};

File: cpp/102_BinaryTreeLevelOrderTraversal.cpp
q./**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
 
//recursive
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        buildVector(root, 0);
        return ret;
    }
private:
    vector<vector<int>> ret;
    
    void buildVector(TreeNode *root, int depth){
        if(root == NULL) return;
        if(ret.size() == depth) ret.push_back(vector<int>());
        ret[depth].push_back(root->val);
        buildVector(root->left, depth+1);
        buildVector(root->right, depth+1);
    }
};

//iterative
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ret;
        if(!root) return ret;
        queue<TreeNode*> q;
        q.push(root);
        q.push(NULL);
        vector<int> cur;
        while(!q.empty()){
            auto t = q.front();
            q.pop();
            if( t == NULL ){
                ret.push_back(cur);
                cur.resize(0);
                if(q.size()) q.push(NULL);
            }else{
                cur.push_back(t->val);
                if(t->left) q.push(t->left);
                if(t->right) q.push(t->right);
            }
        }
        return ret;
    }
};